### DISCUSSION TITLE
Is there a way to detect a measure with a whole rest?

### DISCUSSION THREAD

**Discussion by bearcatjamboree** [mido/mido#524] [Q&A] [answered]
I'm using Python and trying the following to get a list of indexes that have whole rests:

```
def is_whole_rest(track, event_index, whole_rest_ticks):
    # Check if the event at the given index is a whole rest.
    # You may need to adjust this logic depending on your MIDI file structure.
    event = track[event_index]
    return (
        isinstance(event, Message)
        and event.type == "note_off"             # Check if note is a rest
        and event.time >= whole_rest_ticks  # Check if rest is four beats
    )

whole_rest_beats = mid.ticks_per_beat * 4
whole_rest_indices = [i for i in range(len(track)) if is_whole_rest(track, i, whole_rest_beats)]
```

But I'm not getting it to identify measures with rests. Any ideas?  Basically, I just want to make clips from a large file with many, many empty measure that are filled with a single whole rest in 4/4 time.

**Comment by rdoursenaud** [comment, marked as answer]
Hi and welcome!

The short answer: no, not directly.

The long answer:
First, `note_off` does not represent a rest but the time at which the corresponding `note_on` shall stop sounding.
Furthermore, when reading from a MIDI file, the `time` attribute is a delta-time expressed in ticks from the previous event. You can't be certain that the previous event is the `note_on`, it can be any event from any channel or track (depending on the MIDI file `format` aka `type`) including meta-events. Plus sometimes note off is expressed using a `note_on` with its velocity set to `0` to take advantage of the running status feature of the protocol.
So, in summary, rests don't exist per-se in MIDI files and you have to be a bit more clever to detect what might be represented as such in a DAW.

**Reply by rdoursenaud** [reply]
Since what you seem to be searching is the start of a new section (after some rests), I'd rather look for any event with `time` above a certain threshold. This should give you the index of the first event of the next section.

**Reply by bearcatjamboree** [reply]
That won't really do what I want, so I guess I'll just have to scrap the idea.  Thanks for replying though...

**Reply by rdoursenaud** [reply]
You're welcome!

Feel free to elaborate on what you're trying to achieve and maybe provide the file so we can help you.

**Reply by bearcatjamboree** [reply]
Basically, I have a big midi file with lots of "riffs" that I want to extract as separate files and basically eliminating trying to handle them one-by-one.  I was hoping it would be simple splitting the files where I encounter a whole rest, but evidently it is more involved than that.
